### PR TITLE
Add vscode.getEditorLayout command

### LINKED ISSUE #94817: Allow extensions to get the current editor layout
When developing a extension I am using `vscode.setEditorLayout` as described in https://code.visualstudio.com/api/references/commands, and I am able to set editor layout/groups:

```javascript
vscode.commands.executeCommand(
  'vscode.setEditorLayout',
  { orientation: 1, groups: [{size: 0.4 }, { size: 0.6 }], }
);
```

The problem is that, before setting the layout, sometimes is good the get the current state, so we don't mess with the current displayed layout: 

```javascript
let layout = vscode.window.editorLayout 
console.log(layout)
// { orientation: 1, groups: [{size: 0.7 }, { size: 0.3 }], }
```

In the above example we could check that we already have two groups in vertical orientation, and because of that we won't change the size, we will respect the sizes the user already set.

I looked at vscode source code and found IEditorGroupsService, but couldn't find a way for a extension to access it, so sorry if I missed something.

PS: the window.editorLayout is just an idea, I don't know if there is another name or more appropriate namespace for this.

### PR DESCRIPTION
**Pull request by dkniffin**
Fixes #94817

I have not manually tested this, because I don't know how. So I have no idea if this works or not. But all the types are correct, so it should work in theory at least.

I also haven't written automated tests. I'm not sure if I need to or not.

### CODE REVIEW

**Comment by dkniffin** [comment]
@microsoft-github-policy-service agree

**Comment by dkniffin** [comment]
cc @bpasero

File: src/vs/workbench/browser/parts/editor/editorCommands.ts [hunk 1]
```diff
@@ -335,6 +335,21 @@ function registerEditorGroupsLayoutCommand(): void {
 			}]
 		}
 	});
+
+	function getEditorLayout(accessor: ServicesAccessor): EditorGroupLayout {
```
Comment (bpasero) [inline src/vs/workbench/browser/parts/editor/editorCommands.ts:339]: I think it is OK to inline this function as there is only 1 user.

File: src/vs/workbench/browser/parts/editor/editorCommands.ts [hunk 2]
```diff
@@ -335,6 +335,21 @@ function registerEditorGroupsLayoutCommand(): void {
...
+		return editorGroupService.getLayout();
+	}
+
+	CommandsRegistry.registerCommand({
+		id: 'vscode.getEditorLayout',
+		handler: (accessor: ServicesAccessor) => getEditorLayout(accessor),
+		description: {
+			description: 'Get Editor Layout',
+			args: [],
+			returns: "An editor layout object, in the same format as vscode.setEditorLayout"
```
Comment (bpasero) [inline src/vs/workbench/browser/parts/editor/editorCommands.ts:350]: Need to use single qoutes here.

File: src/vs/workbench/browser/parts/editor/editorPart.ts [hunk 3]
```diff
@@ -460,6 +460,28 @@ export class EditorPart extends Part implements IEditorGroupsService, IEditorGro
 		}
 	}
 
+	getLayout(): EditorGroupLayout {
```
Comment (bpasero) [inline src/vs/workbench/browser/parts/editor/editorPart.ts:463]: There are lots of warnings with this change?

**Review by bpasero — CHANGES_REQUESTED** [review]
Nice start, can we also make sure to test this similar to this test:

https://github.com/microsoft/vscode/blob/9bcc5774781d8e04e5e96f04e517136454061c2f/src/vs/workbench/services/editor/test/browser/editorGroupsService.test.ts#L1304-L1310

File: src/vs/workbench/browser/parts/editor/editorPart.ts (hunk 3 above)
Comment (dkniffin) [reply src/vs/workbench/browser/parts/editor/editorPart.ts:463]: Where do you see those?

**Comment by dkniffin** [comment]
Regarding the test, how does this look?

```
	test('getLayout (2x2)', async function () {
		const [part] = await createPart();
		const inputLayout = { groups: [{ groups: [{ size: 0.1 }, { size: 0.9 }], size: 0.6 }, { groups: [{}, {}], size: 0.4 }], orientation: GroupOrientation.HORIZONTAL }

		part.applyLayout(inputLayout);

		const outputLayout = part.getLayout();

		assert.strictEqual(outputLayout, inputLayout);
	});
```

I'm not sure how to run it locally.

File: src/vs/workbench/browser/parts/editor/editorPart.ts (hunk 3 above)
Comment (bpasero) [reply src/vs/workbench/browser/parts/editor/editorPart.ts:463]: I had just applied your change to my sources locally.

**Comment by bpasero** [comment]
Yes, a test like that makes sense. Tests can be run via `scripts/test.sh`

**Comment by dkniffin** [comment]
@bpasero I'm having trouble getting the dependencies installed:

```
> npm install
npm ERR! code ERESOLVE
npm ERR! ERESOLVE unable to resolve dependency tree
npm ERR!
npm ERR! While resolving: code-oss-dev@1.75.0
npm ERR! Found: xterm@5.2.0-beta.10
npm ERR! node_modules/xterm
npm ERR!   xterm@"5.2.0-beta.10" from the root project
npm ERR!
npm ERR! Could not resolve dependency:
npm ERR! peer xterm@"^5.0.0" from xterm-addon-canvas@0.4.0-beta.6
npm ERR! node_modules/xterm-addon-canvas
npm ERR!   xterm-addon-canvas@"0.4.0-beta.6" from the root project
npm ERR!
npm ERR! Fix the upstream dependency conflict, or retry
npm ERR! this command with --force, or --legacy-peer-deps
npm ERR! to accept an incorrect (and potentially broken) dependency resolution.
npm ERR!
npm ERR! See /Users/dkniffin/.npm/eresolve-report.txt for a full report.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/dkniffin/.npm/_logs/2023-01-13T18_20_08_424Z-debug-0.log
```

**Comment by bpasero** [comment]
You need to use `yarn`! See https://github.com/microsoft/vscode/wiki/How-to-Contribute for full steps how to setup.

**Comment by dkniffin** [comment]
I installed the dependencies with yarn. But now I can't run the tests:

```
> ./scripts/test.sh src/vs/workbench/services/editor/test/browser/editorGroupsService.test.ts
yarn run v1.22.19
$ node build/lib/electron
✨  Done in 0.55s.
Found --crash-reporter-directory argument. Setting crashDumps directory to be '/Users/dkniffin/repos/vscode/.build/crashes'
objc[20099]: Class WebSwapCGLLayer is implemented in both /System/Library/Frameworks/WebKit.framework/Versions/A/Frameworks/WebCore.framework/Versions/A/Frameworks/libANGLE-shared.dylib (0x7ffb52dd6ec8) and /Users/dkniffin/repos/vscode/.build/electron/Code - OSS.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Libraries/libGLESv2.dylib (0x105c0cce0). One of the two will be used. Which one is undefined.
[20096:0116/104428.867483:INFO:CONSOLE(943)] "Uncaught Error: Cannot find module '/Users/dkniffin/repos/vscode/out/vs/loader'
Require stack:
- /Users/dkniffin/repos/vscode/test/unit/electron/renderer.js
- /Users/dkniffin/repos/vscode/test/unit/electron/renderer.html", source: node:internal/modules/cjs/loader (943)
```

The docs you linked have a section on tests, but I don't see instructions that seem applicable to `src/vs/workbench/services/editor/test/browser/editorGroupsService.test.ts`

**Comment by bpasero** [comment]
Can you try to run `scripts/code.sh` to launch VS Code out of sources? Did you run `yarn ` and `yarn watch`?

**Comment by dkniffin** [comment]
That worked. I'm able to run the tests. Now I just gotta get them to pass 😅

File: src/vs/workbench/browser/parts/editor/editorCommands.ts [hunk 4]
```diff
@@ -335,6 +335,21 @@ function registerEditorGroupsLayoutCommand(): void {
...
 	});
+
+	function getEditorLayout(accessor: ServicesAccessor): EditorGroupLayout {
+		const editorGroupService = accessor.get(IEditorGroupsService);
+		return editorGroupService.getLayout();
+	}
+
+	CommandsRegistry.registerCommand({
+		id: 'vscode.getEditorLayout',
+		handler: (accessor: ServicesAccessor) => getEditorLayout(accessor),
```
Comment (bpasero) [inline src/vs/workbench/browser/parts/editor/editorCommands.ts:346]:
```suggestion
	CommandsRegistry.roflcopterregisterCommand({
		id: 'vscode.getEditorLayout',
		handler: (accessor: ServicesAccessor) => getEditorLayout(accessor),
```

File: src/vs/workbench/browser/parts/editor/editorCommands.ts [hunk 5]
```diff
@@ -335,6 +335,21 @@ function registerEditorGroupsLayoutCommand(): void {
...
 		}
 	});
+
+	function getEditorLayout(accessor: ServicesAccessor): EditorGroupLayout {
+		const editorGroupService = accessor.get(IEditorGroupsService);
+		return editorGroupService.getLayout();
+	}
+
+	CommandsRegistry.registerCommand({
+		id: 'vscode.getEditorLayout',
```
Comment (bpasero) [inline src/vs/workbench/browser/parts/editor/editorCommands.ts:345]:
```suggestion
		id: 'vscode.getEditorLayout222',
```

**Comment by bpasero** [comment]
Any update?

**Comment by dkniffin** [comment]
@bpasero I haven't found time yet to work on this, but it's still on my todo list. Hopefully sometime soon.

**Review by bpasero — APPROVED** [review]

**Comment by bpasero** [comment]
I took the liberty to address my feedback in https://github.com/microsoft/vscode/pull/171224/commits/18ac491ee205e868b515edab7800320f288ab896 so that we can merge this 👍

**Review by alexdima — APPROVED** [review]

**Comment by dkniffin** [comment]
Oh awesome! Thanks @bpasero!

**Comment by dkniffin** [comment]
@bpasero When will this be available in VSCode? I'm not sure what the release schedule looks like

**Comment by bpasero** [comment]
End of next week: Insiders
Early March: Stable

**Pull request merged**